Implement switching nodes while streaming

alansartorio opened the pull request:
I split the “node-shared” message into “sharing-started” and “node-changed” messages, so that you can start the program once and then you can change the node while it runs.

I used Unix socks to notify the script each time the selected node changes.

Each time the selected node changes I kill the process that connects new nodes and destroy all the links connected to the virtual node, then I run the script that connects new nodes again with the new target node.

I had to move the connection code to a separate script to be able to cleanup after it is killed.

I noticed `pw-dump` sometimes outputs multiple objects instead of one, causing `jq` to produce wrong output. That's why I added `flatten(1)` and `-s` to every jq that parses pw-dump.

<details>
  <summary>How to replicate pw-dump giving extra output</summary>

Leave this running in a terminal:
```sh
while true; do
    pw-dump > /dev/null
done
```
and at the same time run `pw-dump` on another terminal.
The output you'll get is something like:
```
[
  {
    // Expected output
  }
]
[
  {
    "id": 100,
    "info": null
  }
]
[
  {
    "id": 100,
    "info": null
  }
]
[
  {
    "id": 100,
    "info": null
  }
]
```
</details>

alansartorio commented:
I didn't change too much the extension popup code, it probably needs to be changed because you have to scroll to see the "stop" button

IceDBorn commented:
Don't you think that it would be better to remove the stop button entirely and use the share button for both purposes(Start/Stop)?

alansartorio commented:
Yeah, I think that it makes sense

jim3692 commented:
I would prefer to use `$XDG_RUNTIME_DIR` instead of `/tmp` for the socket, as it could conflict in multiuser scenarios.

We should have been using that instead of `/tmp` for security reasons as well.

Another thing is, could we use `mkfifo` instead of sockets? I am not sure which approach is better.

alansartorio commented:
I just implemented the suggestions you made.

- Changed from using `/tmp` to `$XDG_RUNTIME_DIR`.
- Changed from using unix sockets to FIFO.
- Appended the `micId` to the FIFO name to make it unique.

Feel free to suggest more changes

IceDBorn commented:
> Feel free to suggest more changes
Would it be difficult for you to work on the extension part I mentioned above? I could just create a PR on your fork, but I'm on vacation right now and it is kind of impossible for me to code.

alansartorio commented:
Sure!
I don't think I understood what you meant by using the same button for start/stop. Wouldn't it look the same after the change?

IceDBorn commented:
> Sure! I don't think I understood what you meant by using the same button for start/stop. Wouldn't it look the same after the change?

Change it's style and event listeners using JavaScript.

alansartorio commented:
Yeah, I get that. But why would we use one and change the style instead of two? It seems more complicated to me

IceDBorn commented:
> Yeah, I get that. But why would we use one and change the style instead of two separate buttons? It seems more complicated to me EDIT: I just centered the content and removed some margins because it didn't fit when the message, dropdown and button were visible.

I find it easier managing one component with pure JS and HTML. That's why we're going to move to Vue.JS, once I'm back from vacation.

alansartorio commented:
Okay, I'll make it right now.
I support the change to Vue btw

IceDBorn commented:
Is it ready for a review?

alansartorio commented:
There's still an issue that can be a bit annoying. The popup doesn't remember which node you have selected, so if you choose the second node in the dropdown, close the popup and then open it again, the dropdown will reset and show that the first node is selected (the "All desktop audio" one).

IceDBorn commented:
You're free to code!

alansartorio commented:
Done!
I fixed a small issue I noticed while trying to blacklist a node.
And also fixed the issue I was talking about in the previous comment.

You're free to review! 😬

jim3692 commented:
I was testing your implementation and I saw that it can create 1-4 pipewire-screenaudio nodes when Ι start it, but only one of those nodes gets connected and gets removed when I stop. I tried some things but I couldn't figure out the cause of it.

![Untitled](https://github.com/IceDBorn/pipewire-screenaudio/assets/31220180/2ca01a9f-ef12-486a-8997-fe8093b9b8f8)

alansartorio commented:
That's true, it happens to me too.
I noticed the problem comes from the extension part, it probably happened when I unified the two buttons into a single one.
The onclick event listener gets called multiple times. I probably did something wrong when removing the previous event listener.

I just joined the matrix server.

alansartorio commented:
It's fixed now

IceDBorn reviewed (approved):
LGTM :)
